lnwire/features.go

```diff
@@ -0,0 +1,175 @@
+package lnwire
+
+import (
+	"github.com/go-errors/errors"
+	"math"
+	"sync/atomic"
+)
+
+// featureFlag represent the status of the feature optional/required and needed
+// to allows future incompatible changes, or backward compatible changes.
+type featureFlag uint8
+
+// featureName represent the name of the feature and needed in order to have the
+// compile errors if we specify wrong feature name.
+type featureName string
+
+type NameToIndex map[featureName]int
+type indexToFlag map[int]featureFlag
+
+const (
+	// optionalFlag represent the feature which we might be used already but
+	// it isn't required yet, and if remote peer doesn't have this feature
+	// we may turn it off without disconnecting with peer.
+	optionalFlag featureFlag = 2 // 0b10
+
+	// requiredFlag represent the features which is required for proper
+	// peer interaction, we disconnect with peer if it doesn't have this
+	// particular feature.
+	requiredFlag featureFlag = 3 // 0b11
+)
+
+// FeaturesVector represents the global/local feature vector. With this structure
+// you may set/get the feature by name and compare feature vector with remote
+// one.
+type FeaturesVector struct {
+	indexes  NameToIndex
+	features indexToFlag
+	compared uint32
+}
+
+// NewFeatureVector creates new feature vector.
+func NewFeatureVector(indexes NameToIndex) *FeaturesVector {
+	return &FeaturesVector{
+		indexes:  indexes,
+		features: make(indexToFlag),
+	}
+}
+
+// SetFeatureFlag assign flag to the feature.
+func (vector *FeaturesVector) SetFeatureFlag(name featureName,
+	flag featureFlag) error {
+
+	position, ok := vector.indexes[name]
+	if !ok {
+		return errors.Errorf("can't find feature with name: %v", name)
+	}
+
+	vector.features[position] = flag
+	return nil
+}
+
+// IsActive checks - is feature active or not, it might be disable during
+// comparision with remote feature vector if it was optional before and
+// remote peer doesn't support it.
+func (vector *FeaturesVector) IsActive(name featureName) (bool, error) {
+	if atomic.LoadUint32(&vector.compared) == 0 {
```

Comment:
The usage of the atomic variable here is made in an attempt to be thread safe, but the reset of the accessor and mutation methods aren't themselves.